src/common/btree.c

```diff
@@ -4640,3 +4683,191 @@ dbtree_overhead_get(int alloc_overhead, unsigned int tclass, uint64_t otype,
+
+	unknown_flags = nd->tn_flags & ~(BTR_NODE_LEAF | BTR_NODE_ROOT);
+	if (unknown_flags != 0) {
+		report_fn(report_arg, BTR_REPORT_ERROR,
+			  CK_BTREE_NODE_MALFORMED_STR "unknown flags (%#" PRIx16 ")",
+			  unknown_flags);
+		return -DER_NOTYPE;
+	}
+
+	if (nd->tn_pad_32 != 0) {
+		if (error_on_non_zero_padding) {
+			report_fn(report_arg, BTR_REPORT_ERROR,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_PADDING_FMT, nd_off,
+				  nd->tn_pad_32);
+			return -DER_NOTYPE;
+		} else {
+			report_fn(report_arg, BTR_REPORT_WARNING,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_PADDING_FMT, nd_off,
+				  nd->tn_pad_32);
+		}
+	}
+
+	if (nd->tn_gen != 0) {
+		if (error_on_non_zero_padding) {
+			report_fn(report_arg, BTR_REPORT_ERROR,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_GEN_FMT, nd_off, nd->tn_gen);
+			return -DER_NOTYPE;
+		} else {
+			report_fn(report_arg, BTR_REPORT_WARNING,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_GEN_FMT, nd_off, nd->tn_gen);
+		}
+	}
+
+	report_fn(report_arg, BTR_REPORT_MSG, CK_BTREE_NODE_FMT OK_STR, nd_off);
+
+	return DER_SUCCESS;
+}
+
+/**
+ * \struct node_info
+ *
+ * List of node offsets.
+ */
+struct node_info {
+	d_list_t   link;
+	umem_off_t nd_off;
+};
+
+/**
+ * Validate the integrity of a btree.
+ *
+ * \param[in] tcx		Btree context.
+ * \param[in] ck		Checker.
+ *
+ * \retval DER_SUCCESS		The tree is correct.
+ * \retval -DER_NOTYPE		The tree is malformed.
+ * \retval -DER_NONEXIST	The tree is malformed.
+ * \retval -DER_*		Possibly other errors.
+ */
+static int
+btr_nodes_check(struct btr_context *tcx, btr_report_fn_t report_fn, void *report_arg,
+		bool error_on_non_zero_padding)
+{
+	D_LIST_HEAD(node_list);
+	D_LIST_HEAD(child_list);
+	struct node_info *ni;
+	struct node_info *ni_tmp;
+	umem_off_t        nd_off;
+	struct btr_node  *nd;
+	int               rc = DER_SUCCESS;
+
+	D_ASSERT(report_fn != NULL);
+
+	if (btr_root_empty(tcx)) {
+		report_fn(report_arg, BTR_REPORT_MSG, "Empty tree\n");
+		return DER_SUCCESS;
+	}
+
+	D_ASSERT(!btr_has_embedded_value(tcx));
+
+	/** add the root node to the node list */
+	D_ALLOC_PTR(ni);
+	ni->nd_off = tcx->tc_tins.ti_root->tr_node;
+	d_list_add_tail(&ni->link, &node_list);
+
+	/** process the node list */
+	do {
+		d_list_for_each_entry_safe(ni, ni_tmp, &node_list, link) {
+			/** pick the node from the list's head */
+			nd_off = ni->nd_off;
+			nd     = btr_off2ptr(tcx, nd_off);
+
+			/** check the node */
+			rc = btr_node_check(nd, nd_off, report_fn, report_arg,
+					    error_on_non_zero_padding);
+			if (rc != DER_SUCCESS) {
+				break;
+			}
+
+			/** remove the node from the list */
+			d_list_del(&ni->link);
+			D_FREE(ni);
+
+			/** a leaf has no child nodes */
+			if (btr_node_is_leaf(tcx, nd_off)) {
+				continue;
+			}
+
+			/** append the node's children to the children list */
+			for (int at = 0; at < nd->tn_keyn; ++at) {
+				D_ALLOC_PTR(ni);
+				ni->nd_off = btr_node_child_at(tcx, nd_off, at);
+				d_list_add_tail(&ni->link, &child_list);
+			}
```

Comment:
So it is a breadth-first traversing the whole btree, right? Do we need to consider the resource usage? On some degree, the logic can be simplified, for example:
```
while (!d_list_empty(&node_list)) {
   ni = d_list_pop_entry(&node_list, ...);
   ...
   for (int at = 0; at < nd->tn_keyn; ++at) {
			D_ALLOC_PTR(ni);
			ni->nd_off = btr_node_child_at(tcx, nd_off, at);
			d_list_add(&ni->link, &node_list);
	}
};
```
Then traversing become depth-first without `child_list`. Usually, DAOS btree depth is very limited. That may much reduce the resource usage.

Reply:
Very good idea. Thank you.